pkg/apis/azuredisk/v1alpha1/types.go

```diff
@@ -56,22 +57,56 @@ type AzVolumeList struct {
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="NodeName",type=string,JSONPath=`.spec.nodeName`,description="Name of the Node which this AzVolumeAttachment object is attached to"
+// +kubebuilder:printcolumn:name="UnderlyingVolume",type=string,JSONPath=`.spec.underlyingVolume`,description="Name of the Volume which this AzVolumeAttachment object references"
+// +kubebuilder:printcolumn:name="AttachmentTier",type=string,JSONPath=`.spec.attachmentTier`,description="Indicates if the volume attachment should be primary attachment or not"
+// +kubebuilder:printcolumn:name="AttachmentState",type=string,JSONPath=`.status.attachmentState`,description="Indicates if the volume is currently attached or detached to the specified node"
+// +kubebuilder:printcolumn:name="AttachmentTier",type=string,JSONPath=`.status.attachmentTier`,description="Indicates if the volume attachment is primary attachment or not"
 type AzVolumeAttachment struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	Spec   AzVolumeAttachmentSpec   `json:"spec"`
-	Status AzVolumeAttachmentStatus `json:"status"`
+	Spec   AzVolumeAttachmentSpec    `json:"spec"`
+	Status *AzVolumeAttachmentStatus `json:"status,omitempty"`
 }
 
 // AzVolumeAttachmentSpec is the spec for a AzVolumeAttachment resource
 type AzVolumeAttachmentSpec struct {
-	UnderlyingVolume string `json:"underlyingVolume"`
+	UnderlyingVolume string         `json:"underlyingVolume"`
+	NodeName         string         `json:"nodeName"`
+	AttachmentTier   AttachmentTier `json:"attachmentTier"`
 }
 
+// AttachmentTier indicates if the volume attachment is replica attachment or not
+type AttachmentTier string
+
+const (
+	// Primary indicates that the specified node the volume is attached to is where the pod is currently running on
+	Primary AttachmentTier = "Primary"
+	// Replica indicates that the specified node the volume is attached to is one of replicas that pod can failover to if the primary node fails
+	Replica AttachmentTier = "Replica"
+)
+
+// AttachmentState indicates if the volume is currently attached or detached to the specified node
+type AttachmentState string
+
+const (
+	// AttachRequired indicates that the volume needs to be attached to the specified node
+	AttachRequired AttachmentState = "AttachRequired"
+	// Attached indicates the volume has successfully been attached to the specified node
+	Attached AttachmentState = "Attached"
+	// DetachRequired indicates that the volume needs to be detached from the specified node,
+	// the azvolumeattachment object will be deleted after detach operation is completed
+	DetachRequired AttachmentState = "DetachRequired"
+)
+
 // AzVolumeAttachmentStatus is the status for a AzVolumeAttachment resource
 type AzVolumeAttachmentStatus struct {
-	UnderlyingVolume int32 `json:"underlyingVolume"`
+	AttachmentTier  AttachmentTier  `json:"attachmentTier"`
```

Comment:
Why is this repeated in the `Status` when it is already in the `Spec`? I'd think the `State` would be sufficient information in `Status`.

Reply:
It's added to represent cases where we want to turn replicas into primaries.